### PR TITLE
Issue 5235 - Custom tickets show only allowed users/jobs in the assignee list

### LINKED ISSUE #5235: Custom tickets show only allowed users/jobs in the assignee list
### Description
Product issue: https://github.com/3drepo/3D-Repo-Product-Team/issues/590
ATM the list of users/jobs in a custom issue shows all teamspace/jobs. This means that you can assign users/jobs to a ticket while the user/job doesnt have permissions for that federation/container.  For fixing this we are adding some endpoints to fetch the right data.
would recommend to branch from #5209

### Goals
- [ ] As a admin/collab I want to be able to assign tickets to users that have permissions to access the ticket.
- [ ] As a user I only want to be assigned to  tickets that I have permissions to .

### Tasks
- [ ] Add endpoints to api for fetching the assignee list (mock them if backend is not ready)
- [ ] Add fetchs and integration in the ticket card

### Related Resources
[endpoint specs](https://github.com/3drepo/3drepo.io/issues/5209)

### PR DESCRIPTION
**Pull request by Amantini1997**
This fixes #5235 

#### Description
Each model is linked to some users and jobs. To be linked to a user, it means that that user has at least viewer access. To be linked to job, it means at least of one of the linked user is assigned to that job.

Given this, only linked jobs and users are considered valid for a job/user list. Note, for the `assignee property` (i.e. the one in the ticket top panel), only user with a role higher than viewer are considered valid.

As it is possible to have some users/jobs already assigned to user/job property, the invalid values will be indicated:
- a value that still exists in the teamspace will have a red background (for instance, the user has been demoted to `unassigned` for that model);
- a value that cannot be found in the teamspace will be in different list section ("job or user not found"), with a _ad hoc_ circle popover, and with a red background

If it is a multi-select list, toggling ANY job/user, will remove all the invalid jobs/users from the list

#### Test cases
- Add some user/job to the list, and then unlink them. The user/job data should be available in the popover, but the background should be red.
- Add some user/job to the list and then remove them from the teamspace. The user/job data should be unavailable, the background should be red, and hovering the popover circle should better inform the user. Also, the latter users/jobs should be listed in a separate list section

### CODE REVIEW

**Review by The-Daniel — CHANGES_REQUESTED** [review]

**Review by sanmont3drepo — CHANGES_REQUESTED** [review]

File: frontend/src/v5/store/containers/containers.selectors.ts [hunk 1]
```diff
@@ -71,6 +73,32 @@ export const selectHasViewerAccess = createSelector(
...
+	),
+);
+
+export const selectContainerJobs = createSelector(
+	selectContainerById,
+	selectJobs,
+	(container, jobs) => (container?.jobs || []).reduce(
+		(teamspaceJobs, job) => {
+			const teamspaceJob = jobs.find((j) => j._id === job._id);
+			if (!teamspaceJob) return teamspaceJobs;
```
[the remark this answers is not in the document]
Comment (Amantini1997) [reply frontend/src/v5/store/containers/containers.selectors.ts:95]: What I did was this:
- create a job
- assign a user to that job
- create (or edit) a ticket and select that job
- delete that job

[the remark this answers is not in the document]
Comment (sanmont3drepo) [reply frontend/src/v5/store/containers/containers.selectors.ts:95]: Doesnt make sense. There still shouldnt exist in the container but it should exist in the asignees for that particular ticket.

File: frontend/src/v5/ui/components/shared/popoverCircles/errorPopoverCircle/errorPopover.component.tsx [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+ *
+ *  You should have received a copy of the GNU Affero General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import { AvatarWrapper, Data, Heading, PopoverContainer, Username } from '../userPopoverCircle/userPopover/userPopover.styles';
+import { ErrorCircle } from './errorPopoverCircle.styles';
+import { getJobAbbreviation } from '@/v5/store/jobs/jobs.helpers';
+
+export const ErrorPopover = ({ value, message }) => (
```
[the remark this answers is not in the document]
Comment (Amantini1997) [reply frontend/src/v5/ui/components/shared/popoverCircles/errorPopoverCircle/errorPopover.component.tsx:22]: I'm not sure I get this right, but I assume you are referring to what's inside `UserPopover`.
image

If that's the case, the component was not extracted, nor was the `UserPopover` affected by this change.
The idea is that `UserPopover` is called when we know we are dealing with a user that is no longer a teamspace member. This happens, for instance, when you are you looking at a ticket comment from a user that no longer exists in that teamspace
image

The idea here was that we don't know if we are dealing with a user or a job, so calling the `UserPopover` felt wrong. Also, the message shall be different

File: frontend/src/v5/store/containers/containers.selectors.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (Amantini1997) [reply frontend/src/v5/store/containers/containers.selectors.ts:95]: This is for when, in the viewer, you have already fetched the given container/federation jobs/users, but the jobs/users for that teamspace are still being fetched.
As a result, you get `teamspaceJob = undefined`, which eventually leads to the app crash.

You can test this yourself by:
- deleting line `95`;
- heading to the viewer
- expanding a ticket (so the ticketId is the URL) with populated assignees;
- refreshing the page

[the remark this answers is not in the document]
Comment (Amantini1997) [reply frontend/src/v5/store/containers/containers.selectors.ts:95]: **TLDR**
What might be misleading here is the name `teamspaceJob` inside the `if`, as may you think the condition is true when the job does not exist in the teamspace, but it in reality, that's the case if the job has not been fetched yet.


**Longer explanation**
Right now, there 2 ways of fetching a job:
- **Fetch Model Jobs**: fetches the jobs for a given container/federation, which will only provide the job id (and whether or not it's a viewer using the `excludeViewers` search param);
- **Fetch Teamspace Jobs**: fetches all the jobs for a teamspace, which will provide much more data

The problem is in the viewer, when you try to reach a ticket via URL (including `ticketId` as a search param). In this scenario, the model jobs are fetched before the teamspace jobs.

This means that `teamspaceJob` will always be `undefined` until the teamspace jobs are fetched. This way, instead, the selector will return a list with valid jobs

**Review by The-Daniel — APPROVED** [review]

**Review by sanmont3drepo — APPROVED** [review]

**Pull request merged**